src/pkg/caendr/caendr/services/sql/db.py

```diff
@@ -28,93 +28,265 @@
+        Returns:
+          str: [The desired URL filled in with information for the given species.]
+    '''
+    # Make sure wormbase version is set
+    if not self.wb_ver:
+      logger.warning("E_NOT_SET: 'wb_ver'")
+      raise BadRequestError()
+
+    # Make sure a species name was provided if the URL requires one
+    if (self.url_template_type(db_url_name) == 'specific') and species_name is None:
+      logger.warning(f'URL template "{db_url_name}" requires a species, but none was provided.')
+      raise BadRequestError()
+
+    # Make sure the species name is valid
+    if species_name is not None and species_name not in self.species_list.keys():
+      logger.warning(f'Cannot construct URL for unknown species "{species_name}".')
+      raise BadRequestError()
+
+    # Get the desired template an fill in species information, if applicable
+    t = Template(self.get_url_template(db_url_name))
+    if species_name is not None:
+      return t.substitute({
+        'WB':      self.wb_ver,
+        'SPECIES': species_name,
+        'PRJ':     self.species_list[species_name]['project_number'],
+      })
+    else:
+      return t.substitute({
+        'WB':      self.wb_ver,
+      })
+
+
+  def get_filename(self, db_url_name: str, species_name: str = ''):
+    '''
+      Gets the local filename for a URL template and species.  Does NOT guarantee that this file exists.
+        Args:
+          db_url_name (str): [The name of the template to retrieve.]
+          species_name (str, optional): [The species to fill in the template with. Defaults to None. Optional, but must be provided for certain templates.]
+        Returns:
+          str: [The location in the local downloads folder where the downloaded file belongs.]
+    '''
+    if (self.url_template_type(db_url_name) == 'generic'):
+      url_path = ''
+    else:
+      url_path = species_name + '/'
+    return f'{self.local_download_path}/{url_path}{db_url_name}'
+
+
+  ## Download external databases ##
+
+  def prefetch_all_external_dbs(self):
+    '''
+      Downloads all external DB files and saves them locally.
+    '''
+    logger.info('Downloading All External DBs...')
+
+    # Make sure wormbase version is set
+    if not self.wb_ver:
+      logger.warning("E_NOT_SET: 'wb_ver'")
+      raise BadRequestError()
+
+    # Download all files that depend on species
+    for species_name in species_list.keys():
+      for url_template_name in external_db_url_templates['specific'].keys():
+        self.fetch_external_db(url_template_name, species_name)
+
+    # Download all files that don't depend on species
+    for url_template_name in external_db_url_templates['generic'].keys():
+      self.fetch_external_db(url_template_name)
+
+    logger.info('Done Downloading All External Data.')
+
+
+  def fetch_external_db(self, db_url_name: str, species_name: str = None, force: bool = False):
```

Comment:
consider renaming `force` with `use_cache`

Reply:
Fixed in [70ee6b0](https://github.com/AndersenLab/CAENDR/pull/131/commits/70ee6b0bd37ca68860c5031f93389ebdf94d5f22)